File: code/92.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-07 19:58:48
LastEditors: lmio 2091319361@qq.com
Description: 92. 反转链表 II
'''

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverseBetween(head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
    dummy = ListNode(-1)
    dummy.next = head
    prev = dummy
    for _ in range(left - 1):
        prev = prev.next
    cur = prev.next
    # 不断把next插入到prev后面
    for _ in range(right - left):
        next = cur.next
        cur.next = next.next
        next.next = prev.next
        prev.next = next
    return dummy.next

File: code_cpp/92.cpp
/*
 * @Author: lmio 2091319361@qq.com
 * @Date: 2024-08-22 15:39:45
 * @LastEditors: lmio 2091319361@qq.com
 * @Description: 92. 反转链表 II
 */
#include "utils.h"
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int left, int right) {
        auto dummy = new ListNode(-1, head);
        auto pre = dummy;
        auto cur = head;
        for (int i = 0; i < left - 1; ++i) {
            pre = pre->next;
        }
        cur = pre->next;
        auto last = cur;
        for (int i = 0; i < right - left; ++i) {
            auto tmp = cur->next;
            cur->next = tmp->next;
            tmp->next = pre->next;
            pre->next = tmp;
        }
        return dummy->next;
    }
};